services/gundeck/src/Gundeck/API.hs

```diff
@@ -131,7 +137,94 @@ sitemap = do
 
 type JSON = Media "application" "json"
 
-docs :: ByteString ::: JSON -> Gundeck Response
-docs (url ::: _) =
+docsH :: ByteString ::: JSON -> Gundeck Response
+docsH (url ::: _) =
   let doc = mkSwaggerApi (decodeLatin1 url) Model.gundeckModels sitemap
    in return $ json doc
+
+addTokenH :: UserId ::: ConnId ::: JsonRequest PushToken ::: JSON -> Gundeck Response
+addTokenH (uid ::: cid ::: req ::: _) = do
+  newtok <- fromJsonBody req
+  handleAddTokenResponse <$> Push.addToken uid cid newtok
+
+handleAddTokenResponse :: Push.AddTokenResponse -> Response
+handleAddTokenResponse = \case
+  Push.AddTokenSuccess newtok -> success newtok
+  Push.AddTokenNoBudget -> snsThreadBudgetReached
+  Push.AddTokenNotFound -> notFound
+  Push.AddTokenInvalid -> invalidToken
+  Push.AddTokenTooLong -> tokenTooLong
+  Push.AddTokenMetadataTooLong -> metadataTooLong
+
+success :: PushToken -> Response
+success t =
+  let loc = Text.encodeUtf8 . tokenText $ t ^. token
+   in json t & setStatus status201 & addHeader hLocation loc
+
+invalidToken :: Response
+invalidToken =
+  json (Error status400 "invalid-token" "Invalid push token")
+    & setStatus status404
+
+snsThreadBudgetReached :: Response
+snsThreadBudgetReached =
+  json (Error status400 "sns-thread-budget-reached" "Too many concurrent calls to SNS; is SNS down?")
+    & setStatus status413
+
+tokenTooLong :: Response
+tokenTooLong =
+  json (Error status400 "token-too-long" "Push token length must be < 8192 for GCM or 400 for APNS")
+    & setStatus status413
+
+metadataTooLong :: Response
+metadataTooLong =
+  json (Error status400 "metadata-too-long" "Tried to add token to endpoint resulting in metadata length > 2048")
+    & setStatus status413
+
+notFound :: Response
+notFound = empty & setStatus status404
+
+deleteTokenH :: UserId ::: Token ::: JSON -> Gundeck Response
+deleteTokenH (uid ::: tok ::: _) = setStatus status204 empty <$ Push.deleteToken uid tok
+
+listTokensH :: UserId ::: JSON -> Gundeck Response
+listTokensH (uid ::: _) = setStatus status200 . json <$> Push.listTokens uid
+
+pushH :: Request ::: JSON -> Gundeck Response
+pushH (req ::: _) = do
+  ps <- fromJsonBody (JsonRequest req)
+  empty <$ Push.push ps
+
+paginateH :: JSON ::: UserId ::: Maybe ByteString ::: Maybe ClientId ::: Range 100 10000 Int32 -> Gundeck Response
+paginateH (_ ::: uid ::: sinceRaw ::: clt ::: size) = do
+  Notification.PaginateResult gap page <- Notification.paginate uid (join since) clt size
+  pure . updStatus gap . json $ page
+  where
+    since :: Maybe (Maybe NotificationId)
+    since = parseUUID <$> sinceRaw
+    --
+    parseUUID :: ByteString -> Maybe NotificationId
+    parseUUID = UUID.fromASCIIBytes >=> isV1UUID >=> return . Id
+    --
+    isV1UUID :: UUID -> Maybe UUID
+    isV1UUID u = if UUID.version u == 1 then Just u else Nothing
+    --
+    -- the motivation of updStatus is far from clear to me, but this is how it's always been done.
```

Comment:
I have the same! I was confused about this. I was just reading the docs and figuring out why   we return `404` but have notifications to serve.

Reply:
my theory is that the client wanted something else, so we want to say "too bad", but at the same time we want to deliver something that may still be better than nothing.  would be nice if the author had left a comment about what's going on, though.

Reply:
So there is one case that kind of makes sense to me: When `resultGap` is true (which means "a start ID ('since') has been given which could not be found", usually when a client missed notifications by being offline for too long), we return the `QueuedNotificationList`, but set status 404 to tell the client that something is missing. It's now up to the client to just use those notifications or force a slow sync. This could be a flag in the JSON response instead of a status code, but I see how it's used.

But the other case that confuses me more is that if the `since` query parameter is not a valid V1 UUID, we still fetch notifications and give a 404. So `since` could be any string at all and the result would not be distinguishable from having requested a valid UUID that was just too old. Do we do this for some kind of backwards compatibility reason? Why would that happen, so we can't assume this to be an invalid request?

Before changing this, we should have a discussion with someone who knows the history of this (@tiago-loureiro) or talk to the client teams.

Reply:
I have added to the documentation this weird behaviour.

Given we have an explicit test that asserts this, lets not change it for now. but I agree it's awkward https://github.com/zinfra/orlop/pull/30/commits/a358dfc1cb225c92066ea79db28c8824531ae231


I have a proposal to make this API a bit more elegant 

What if we _include_ the `since` notification in the response list? i.e. it's an inclusive bound instead of an exclusive bound. Then clients
can distinguish by the notification being there and not being there by ... checking that it's not there.